### PR TITLE
E2E test various metric types

### PR DESCRIPTION
**Pull request by skhalash**
## Description

Changes proposed in this pull request:

- Add a new matcher that checks if there is at least one metric that satisfies a given predicate
- Extend the existing Prometheus input test to check not only the metric names but also the type and origin labels. You can find the actual exposed metrics here: https://github.com/kyma-project/examples/blob/main/prometheus/monitoring-custom-metrics/main.go.

## Related Issues and Documents

https://github.com/kyma-project/kyma/issues/17828

## Traceability

- [x] The PR is linked to a GitHub Issue.
- [ ] New features have a milestone label set.
- [x] New features have defined acceptance criteria in a corresponding GitHub Issue, and all criteria are satisfied with this PR.
- [x] The corresponding GitHub Issue has a respective `area` label.
- [ ] The follow-up issues (if any) are linked in the `Related Issues` section.

## Testability

The feature is unit-tested:

- [ ] Yes.
- [x] No, because unit tests are not needed.
- [ ] No, because of ...

The feature is e2e-tested:

- [ ] Yes.
- [x] No, because e2e-tests are not needed.
- [ ] No, because of ...

Tests conducted for the PR:

## Codebase

- [x] My code follows the style guidelines of this project.
- [x] The code was planned and designed following the defined architecture and the separation of concerns.
- [ ] The code has sufficient comments, particularly for all hard-to-understand areas.
- [x] This PR adds value and shows no feature creep.
- [ ] I have augmented the test suite that proves my fix is effective or that my feature works.
- [ ] Adjusted the documentation if the change is user-facing.

### CODE REVIEW

File: test/e2e/testkit/matchers/metric_matchers_test.go [hunk 1]
```diff
@@ -209,6 +209,87 @@ var _ = Describe("HaveSumMetrics", Label("metrics"), func() {
...
+
+	BeforeEach(func() {
+	})
+
+	Context("with nil input", func() {
+		It("should error", func() {
+			success, err := HaveMetricsThatSatisfy(func(metric pmetric.Metric) bool {
+				return true
+			}).Match(nil)
+			Expect(err).Should(HaveOccurred())
```
Comment (ameteiko) [inline test/e2e/testkit/matchers/metric_matchers_test.go:224]: My code-style preference is to separate the Arrange, Act and Assert blocks of the test with empty lines. IMO it improves the readability, but it is a subjective preference.

File: test/e2e/testkit/otlp/metrics/metrics.go [hunk 2]
```diff
@@ -183,6 +183,46 @@ func makeUnique(slice []string) []string {
 	return uniqueList
 }
 
+func AllDataPointsHaveAttributes(m pmetric.Metric, expectedAttrKeys ...string) bool {
+	var attrsPerDataPoint []pcommon.Map
+
+	switch m.Type() {
+	case pmetric.MetricTypeSum:
+		for i := 0; i < m.Sum().DataPoints().Len(); i++ {
+			attrsPerDataPoint = append(attrsPerDataPoint, m.Sum().DataPoints().At(i).Attributes())
```
Comment (ameteiko) [inline test/e2e/testkit/otlp/metrics/metrics.go:192]: Do `m.Sum().DataPoints()`, `m.Gauge().DataPoints()`, etc. return the same instance? If so, maybe we can simplify these 4 loops into just one?

**Review by ameteiko — CHANGES_REQUESTED** [review]

File: test/e2e/metrics_prometheus_input_test.go [hunk 3]
```diff
@@ -70,17 +73,37 @@ var _ = Describe("Metrics Prometheus Input", Label("metrics"), func() {
...
-		It("Should verify custom metric names", func() {
+		It("Should verify custom metrics", func() {
 			Eventually(func(g Gomega) {
 				resp, err := proxyClient.Get(urls.MockBackendExport())
 				g.Expect(err).NotTo(HaveOccurred())
 				g.Expect(resp).To(HaveHTTPStatus(http.StatusOK))
 				g.Expect(resp).To(HaveHTTPBody(SatisfyAll(
-					HaveMetricNames(mocks.CustomMetricNames...))))
+					HaveMetricsThatSatisfy(func(m pmetric.Metric) bool {
+						return m.Name() == "cpu_temperature_celsius" && m.Type() == pmetric.MetricTypeGauge
```
[the remark this answers is not in the document]
Comment (skhalash) [reply test/e2e/metrics_prometheus_input_test.go:83]: Agree, will move the metric names/types/labels closer to the custom metric provider mock definition

File: test/e2e/testkit/matchers/metric_matchers.go [hunk 4]
```diff
@@ -64,6 +64,28 @@ func HaveSumMetrics(expectedMetrics ...pmetric.Metric) types.GomegaMatcher {
 	}, gomega.ContainElements(expectedMetrics))
 }
 
+type MetricPredicate = func(pmetric.Metric) bool
+
+func HaveMetricsThatSatisfy(predicate MetricPredicate) types.GomegaMatcher {
+	return gomega.WithTransform(func(actual interface{}) ([]pmetric.Metric, error) {
+		actualBytes, ok := actual.([]byte)
+		if !ok {
+			return nil, fmt.Errorf("HaveMetrics requires a []byte, but got %T", actual)
```
[the remark this answers is not in the document]
Comment (skhalash) [reply test/e2e/testkit/matchers/metric_matchers.go:73]: Agree, will change

File: test/e2e/metrics_prometheus_input_test.go [hunk 5]
```diff
@@ -10,12 +10,15 @@ import (
 	"k8s.io/apimachinery/pkg/types"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
+	"go.opentelemetry.io/collector/pdata/pmetric"
+
 	kitk8s "github.com/kyma-project/telemetry-manager/test/e2e/testkit/k8s"
 	"github.com/kyma-project/telemetry-manager/test/e2e/testkit/k8s/verifiers"
 	"github.com/kyma-project/telemetry-manager/test/e2e/testkit/kyma"
 	kitmetric "github.com/kyma-project/telemetry-manager/test/e2e/testkit/kyma/telemetry/metric"
```
[the remark this answers is not in the document]
Comment (skhalash) [reply test/e2e/metrics_prometheus_input_test.go:18]: Hm, not sure, I find both aliases not quite descriptive. But I cannot come up with a better one 😄

File: test/e2e/metrics_prometheus_input_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (skhalash) [reply test/e2e/metrics_prometheus_input_test.go]: Done

File: test/e2e/testkit/matchers/metric_matchers.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (skhalash) [reply test/e2e/testkit/matchers/metric_matchers.go:73]: Done

File: test/e2e/testkit/matchers/metric_matchers.go [hunk 6]
```diff
@@ -64,6 +64,28 @@ func HaveSumMetrics(expectedMetrics ...pmetric.Metric) types.GomegaMatcher {
...
+		}
+
+		actualMds, err := unmarshalOTLPJSONMetrics(actualBytes)
```
[the remark this answers is not in the document]
Comment (skhalash) [reply test/e2e/testkit/matchers/metric_matchers.go:76]: `Mds` stands for metric data (s 😄). The actual variable naming was stolen from the collector codebase (for example, [here](https://github.com/open-telemetry/opentelemetry-collector/blob/3089ea878eef022913a053600f3599b3fddffb95/receiver/otlpreceiver/internal/metrics/otlp_test.go#L27)). The problem is that there is `pmetric.Metric` and `pmetric.Metrics`, which makes variable naming harder.

File: test/e2e/testkit/matchers/metric_matchers_test.go [hunk 7]
```diff
@@ -209,6 +209,87 @@ var _ = Describe("HaveSumMetrics", Label("metrics"), func() {
 	})
 })
 
+var _ = Describe("HaveMetricsThatSatisfy", Label("metrics"), func() {
+	var fileBytes []byte
+	//var expectedMetrics []pmetric.Metric
+
+	BeforeEach(func() {
```
[the remark this answers is not in the document]
Comment (skhalash) [reply test/e2e/testkit/matchers/metric_matchers_test.go:216]: Drop

File: test/e2e/testkit/matchers/metric_matchers_test.go (hunk 1 above)
Comment (skhalash) [reply test/e2e/testkit/matchers/metric_matchers_test.go:224]: As discussed, replaced it with a one-liner

**Pull request merged**